addons/web_editor/static/src/js/editor/snippets.options.js

```diff
@@ -4183,23 +4183,56 @@ registry.BackgroundShape = SnippetOptionWidget.extend({
      * Returns the default colors for the currently selected shape.
      *
      * @private
-     */
-    _getDefaultColors() {
-        const $shapeContainer = this.$target.find('> .o_we_shape')
-            .clone()
-            .addClass('d-none')
-            // Needs to be in document for bg-image class to take effect
-            .appendTo(document.body);
-        const shapeContainer = $shapeContainer[0];
-        $shapeContainer.css('background-image', '');
-        const shapeSrc = shapeContainer && getBgImageURL(shapeContainer);
-        $shapeContainer.remove();
+     * @param {String} shape identifier of the shape to consider instead of the currenlty selected
+     */
+    _getDefaultColors(shape) {
+        let shapeSrc;
+        if (shape) {
+            // locate original in snippet options
+            const $shapeContainer = this.$el.find(".o_we_shape_menu we-button[data-shape='" + shape + "'] div.o_we_shape");
+            const shapeContainer = $shapeContainer[0];
+            shapeSrc = shapeContainer && getBgImageURL(shapeContainer);
+        } else {
+            const $shapeContainer = this.$target.find('> .o_we_shape')
+                .clone()
+                .addClass('d-none')
+                // Needs to be in document for bg-image class to take effect
+                .appendTo(document.body);
+            const shapeContainer = $shapeContainer[0];
+            $shapeContainer.css('background-image', '');
+            shapeSrc = shapeContainer && getBgImageURL(shapeContainer);
+            $shapeContainer.remove();
+        }
         if (!shapeSrc) {
             return {};
         }
         const url = new URL(shapeSrc, window.location.origin);
         return Object.fromEntries(url.searchParams.entries());
     },
+    /**
+     * Returns the implicit colors for the currently selected shape.
+     *
+     * The implicit colors are use upon shape selection. They are computed as:
+     * - the default colors
+     * - patched with each set of colors of previous siblings shape
+     * - patched with the colors of the previously selected shape
+     * - filtered to only keep the colors involved in the current shape
+     *
+     * @private
+     * @param {String} shape identifier of the selected shape
+     * @param {Object} previousColors colors of the shape before its replacement
+     */
+    _getImplicitColors(shape, previousColors) {
+        const defaultColors = this._getDefaultColors(shape);
+        let colors = _.defaults({}, previousColors);
+        let sibling = this.$target[0].previousElementSibling;
+        while (sibling) {
+            colors = _.defaults(colors, this._getShapeData(sibling).colors || {});
```

Comment:
`_getShapeData` should always return an object with a filled color field, no? Is that empty object fallback necessary?

Also note that `_.defaults` acts in place so if you do use it, the assignment is redundant (the return value is meant to be used for chaining or uses in other expressions)

You can use `Object.assign` instead of `_.defaults` here though, simply reverse the order of the arguments (`Object.assign(defaults, object)`), you will need an assignment in that case because `Object.assign` acts on the first parameter (if you want to leave the defaults object untouched, which is not the case here, you can use an empty object as the first argument)

Also note that while `_getShapeData` accepts a target as argument, it will return the correct colors only if they're customized, if they're the default colors, it will return the default colors for the *curent shape* because `_getDefaultColors` doesn't accept a target as an argument currently. I don't think it matters in this case since you don't care about colors when they're defaults, but something to keep in mind.

Reply:
Ultimately also made _getShapeData return the correct default colors for the given shape - because wrong defaults might override specified colors coming from above in the siblings chain.